# Series

## Install

pip install pandas

## import

In [27]:
import pandas as pd
import numpy as np

In [28]:
### Index and Data Lists

# We can create a Series from Python lists (also from NumPy arrays)

data = pd.Series([0.25, 0.5, 0.75, 1.0],
    index=['C', 'P', 'S', 'Y'])
data


C    0.25
P    0.50
S    0.75
Y    1.00
dtype: float64

### From a  Dictionary

In [29]:
umur = {'Andi': 20, 'Budi': 30, 'Caca': 40}
umur = pd.Series(umur)
umur

Andi    20
Budi    30
Caca    40
dtype: int64

In [30]:
penjualan_per_orang = {'Andi': 2000000, 'Budi': 3000000, 'Caca': 4000000}
penjualan_per_tim = {'Tim A ': 2000000, 'Tim B': 3000000, 'Tim C': 4000000}


In [31]:
penjualan_per_orang = pd.Series(penjualan_per_orang)
penjualan_per_tim = pd.Series(penjualan_per_tim)
print(penjualan_per_orang)
print(penjualan_per_tim)

Andi    2000000
Budi    3000000
Caca    4000000
dtype: int64
Tim A     2000000
Tim B     3000000
Tim C     4000000
dtype: int64


## Reading a CSV

In [33]:
pwd

'c:\\Users\\junai\\Documents\\Project\\python\\RisetML\\CPS\\02.pandas'

In [35]:
ls

 Volume in drive C is Windows-SSD
 Volume Serial Number is 800B-EF6F

 Directory of c:\Users\junai\Documents\Project\python\RisetML\CPS\02.pandas

18/11/2024  03.41    <DIR>          .
18/11/2024  03.39    <DIR>          ..
18/11/2024  03.16                 0 01.seriesAndDataFrame.ipynb
16/07/2024  10.29            18.750 tips.csv
               2 File(s)         18.750 bytes
               2 Dir(s)  160.685.621.248 bytes free


In [37]:
data = pd.read_csv('tips.csv')
data

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

In [38]:
# display the first 5 rows
data.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [39]:
# display the last 5 rows
data.tail()

total_bill   tip     sex smoker   day    time  size  price_per_person  \
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
239    Michael Avila  5296068606052842    Sat2657  
240   Monica Sanders  3506806155565404    Sat1766  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

----
About this DataSet
* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:
* Format
    * A data frame with 244 rows and 7 variables
* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.
In all he recorded 244 tips. The data was reported in a collection of case studies for business statistics (Bryant & Smith 1995).
* References
    * Bryant, P. G. and Smith, M (1995) Practical Data Analysis: Case Studies in Business Statistics. Homewood, IL: Richard D. Irwin Publishing:
    
* Note: We created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

In [40]:
# display the data types of each column
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


In [41]:
# display the summary statistics of numeric columns
data.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

In [42]:
# select a single column
data['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [43]:
# select multiple columns
data[['total_bill', 'tip', 'sex']]

total_bill   tip     sex
0         16.99  1.01  Female
1         10.34  1.66    Male
2         21.01  3.50    Male
3         23.68  3.31    Male
4         24.59  3.61  Female
..          ...   ...     ...
239       29.03  5.92    Male
240       27.18  2.00  Female
241       22.67  2.00    Male
242       17.82  1.75    Male
243       18.78  3.00  Female

[244 rows x 3 columns]

In [44]:
# create a new column
data['total_per_person'] = data['total_bill'] / data['size']
data.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  total_per_person  
0  Christy Cunningham  3560325168603410    Sun2959          8.495000  
1      Douglas Tucker  4478071379779230    Sun4608          3.446667  
2      Travis Walters  6011812112971322    Sun4458          7.003333  
3    Nathaniel Harris  4676137647685994    Sun5260         11.840000  
4        Tonya Carter  4832732618637221    Sun2251          6.147500

In [45]:
# delete a column
del data['total_per_person']
data.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [47]:
data.set_index('time')

total_bill   tip     sex smoker   day  size  price_per_person  \
time                                                                    
Dinner       16.99  1.01  Female     No   Sun     2              8.49   
Dinner       10.34  1.66    Male     No   Sun     3              3.45   
Dinner       21.01  3.50    Male     No   Sun     3              7.00   
Dinner       23.68  3.31    Male     No   Sun     2             11.84   
Dinner       24.59  3.61  Female     No   Sun     4              6.15   
...            ...   ...     ...    ...   ...   ...               ...   
Dinner       29.03  5.92    Male     No   Sat     3              9.68   
Dinner       27.18  2.00  Female    Yes   Sat     2             13.59   
Dinner       22.67  2.00    Male    Yes   Sat     2             11.34   
Dinner       17.82  1.75    Male     No   Sat     2              8.91   
Dinner       18.78  3.00  Female     No  Thur     2              9.39   

                Payer Name         CC Number Payment ID  
time                                                     
Dinner  Christy Cunningham  3560325168603410    Sun2959  
Dinner      Douglas Tucker  4478071379779230    Sun4608  
Dinner      Travis Walters  6011812112971322    Sun4458  
Dinner    Nathaniel Harris  4676137647685994    Sun5260  
Dinner        Tonya Carter  4832732618637221    Sun2251  
...                    ...               ...        ...  
Dinner       Michael Avila  5296068606052842    Sat2657  
Dinner      Monica Sanders  3506806155565404    Sat1766  
Dinner          Keith Wong  6011891618747196    Sat3880  
Dinner        Dennis Dixon     4375220550950      Sat17  
Dinner     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 10 columns]